qui/tray/devices.py

```diff
@@ -256,56 +232,84 @@ def __init__(self, menu: Gtk.Menu, device_manager):
             self.positions[dev_type] = pos
             self.separators[dev_type] = separator
 
-        self.device_manager.connect_to_signal("Added", self.add)
-        self.device_manager.connect_to_signal("Removed", self.remove)
-
-    def add(self, dev_obj_path: dbus.ObjectPath):
-        dev = self.device_manager[dev_obj_path]
-        if dev['dev_class'] not in DEV_TYPES:
+        for devclass in DEV_TYPES:
+            self.dispatcher.add_handler('device-attach:' + devclass,
+                                        self.device_attached)
+            self.dispatcher.add_handler('device-detach:' + devclass,
+                                        self.device_detached)
+            self.dispatcher.add_handler('device-list-change:' + devclass,
+                                        self.device_change)
+
+    def update_device_list(self):
+        devices = {}
+        for vm in self.qapp.domains:
+            for devclass in DEV_TYPES:
+                for device in vm.devices[devclass]:
+                    devices[device] = []
+
+        for vm in self.qapp.domains:
+            for devclass in DEV_TYPES:
+                for device in vm.devices[devclass].attached():
+                    if device in devices:
+                        # occassionally ghost UnknownDevices appear when a
+                        # device was removed but not detached from a VM
+                        devices[device].append(vm)
+
+        for device in [dev for dev in devices if dev not in self.menu_items.keys()]:
+            self.add(device, devices[device])
+
+        for device in [dev for dev in self.menu_items.keys() if dev not in devices]:
+            self.remove(device)
+
+    def device_change(self, _vm, _event, **_kwargs):
+        self.update_device_list()
+
+    def add(self, device, frontend_domains):
+        if device.devclass not in DEV_TYPES:
             return
 
-        position = self._position(dev['dev_class'])
+        name = device.backend_domain.name + ':' + device.ident
+        position = self._position(device.devclass)
 
-        #sort function
-        name = dev.backend_domain['name'] + ':' + dev['ident']
-        for i in self.menu_items:
-            if str(dev['dev_class']) == str(i.dev['dev_class']):
-                if str(i.dev.backend_domain['name'] + ':' + i.dev['ident']) <= str(name):
-                    position += 1
-        self._insert(dev_obj_path, position)
+        position += len(
+            [dev for dev in self.menu_items.keys()
+             if dev.devclass == device.devclass
+             and (dev.backend_domain.name + ':' + dev.ident) < name])
 
-        if dev['dev_class'] not in [DEV_TYPES[0], DEV_TYPES[-1]]:
-            self.separators[dev['dev_class']].show()
+        self._insert(device, frontend_domains, position)
 
-        subprocess.call(['notify-send', "Device %s is available" % (dev.name)])
+        if device.devclass not in [DEV_TYPES[0], DEV_TYPES[-1]]:
```

Comment:
This condition looks wrong - if there are only devices from two dev classes, there will be no separator.
If I'm wrong, it worth adding a comment where is it handled.